src/eth/relayer/transaction_dag.rs

```diff
@@ -128,18 +128,48 @@ impl TransactionDag {
         }
     }
 
-    fn add_nonce_edges(dag: &mut StableGraph<TransactionMined, i32>, node_indexes: &HashMap<(BlockNumber, Index), NodeIndex>) {
-        for (i, (tx1_id, tx1_idx)) in node_indexes.iter().sorted_by_key(|(idx, _)| **idx).enumerate() {
-            let tx1 = dag.node_weight(*tx1_idx).unwrap();
-            let (tx1_signer, tx1_nonce) = (tx1.input.signer, tx1.input.nonce);
-            for (tx2_id, tx2_idx) in node_indexes.iter().sorted_by_key(|(idx, _)| **idx).skip(i + 1) {
-                let tx2 = dag.node_weight(*tx2_idx).unwrap();
-                let (tx2_signer, tx2_nonce) = (tx2.input.signer, tx2.input.nonce);
-                if tx1_signer == tx2_signer && tx1_nonce.next() == tx2_nonce {
-                    dag.add_edge(*node_indexes.get(tx1_id).unwrap(), *node_indexes.get(tx2_id).unwrap(), 1);
+    pub fn split_components(self) -> Vec<Self> {
+        let mut dag = self.dag;
+        let mut ret = vec![];
+
+        while dag.node_count() > 0 {
+            let mut root = None;
+
+            // find a root
+            for index in dag.node_identifiers() {
+                if dag.neighbors_directed(index, petgraph::Direction::Incoming).next().is_none() {
+                    root = Some(index);
+                    break;
                 }
             }
```

Comment:
lil suggestion for no mutability
```rust
let root = dag
    .node_identifiers()
    .position(|index| dag.neighbors_directed(index, Incoming).next().is_none());
```

---

instead of running this search multiple times inside of this `while dag.node_count() > 0 {` loop, I wonder if you can't just run it once, grab all roots, and then calculate the value of `ret` for each one of them.

Reply:
I could run it only once, will improve in the next PR.